### PR TITLE
Requesting merge of QuadVtx algorithm

### PR DESCRIPTION
**Pull request by cjbacchus**
QuadVtx is a vertex finder with competitive performance to the Pandora vertex, achieved with a simple and fast algorithm with no dependencies (it starts from the hits).

The most recent presentation of this algorithm was
https://indico.fnal.gov/event/21520/contribution/0

Apologies if this is misformatted in some way, I'm experimenting with the new github-based system.

### CODE REVIEW

File: larreco/QuadVtx/EvalVtx_module.cc
Comment (knoepfel) [inline larreco/QuadVtx/EvalVtx_module.cc]: If a user wants to configure two instances of the `EvalVtx`, then these global variables will be shared between the two module instances and the `TTree` constructed in one of the modules can easily be malformed.  Please make the variables members of the `EvalVtx` class to avoid this problem.

File: larreco/QuadVtx/QuadVtx_module.cc
Comment (knoepfel) [inline larreco/QuadVtx/QuadVtx_module.cc]: Suggest `constexpr size_t` instead of `const size_t`.

File: larreco/QuadVtx/QuadVtx_module.cc
Comment (knoepfel) [inline larreco/QuadVtx/QuadVtx_module.cc]: Suggest replacing hard-coded value of pi with `M_PI`.  With C++20, there will be an actual pi variable defined by the standard.

File: larreco/QuadVtx/QuadVtx_module.cc
Comment (knoepfel) [inline larreco/QuadVtx/QuadVtx_module.cc]: `std::random_shuffle` has been removed from C++17; please replace with something else (e.g. `std::shuffle`).

File: larreco/QuadVtx/QuadVtx_module.cc
Comment (knoepfel) [inline larreco/QuadVtx/QuadVtx_module.cc]: Recommend moving this to just before the `MapFromLines` call.

File: larreco/QuadVtx/QuadVtx_module.cc
Comment (knoepfel) [inline larreco/QuadVtx/QuadVtx_module.cc]: Recommend either making the `min/max` variables of type `double`, or changing `Pt2D::x` and `Pt2D::z` to be of type `float`.

File: larreco/QuadVtx/QuadVtx_module.cc
Comment (knoepfel) [inline larreco/QuadVtx/QuadVtx_module.cc]: Two issues:

- At the very least, use a `std::unique_ptr` instead of a bare pointer for exception safety reasons.
- As `evt_dir` is not actually used in this scope, I suggest moving it to `FindVtx` and changing the call to be:

```cpp
if(FindVtx(*hits, vtx, evt.event())) {
  ...
}
```
thus moving the service-handle creation to the `FindVtx` function.

**Review by knoepfel — CHANGES_REQUESTED** [review]

File: larreco/QuadVtx/QuadVtx_module.cc
Comment (knoepfel) [inline larreco/QuadVtx/QuadVtx_module.cc]: Recommend `constexpr size_t`.

File: larreco/QuadVtx/QuadVtx_module.cc
Comment (knoepfel) [inline larreco/QuadVtx/QuadVtx_module.cc]: There are various free functions in this file that should be placed in a namespace to keep them invisible to other parts of the program--e.g. `IntersectsCircle`, `LinesFromPoints`, `FindPeak3D`, etc.

File: larreco/QuadVtx/QuadVtx_module.cc
Comment (knoepfel) [inline larreco/QuadVtx/QuadVtx_module.cc]: Change to `private`.

File: larreco/QuadVtx/HeatMap.h
Comment (knoepfel) [inline larreco/QuadVtx/HeatMap.h]: Should be `private`.

File: larreco/QuadVtx/HeatMap.h
Comment (knoepfel) [inline larreco/QuadVtx/HeatMap.h]: Can these data members be made private?  Not generally a good idea to expose member data of classes.

File: larreco/QuadVtx/HeatMap.h
Comment (knoepfel) [inline larreco/QuadVtx/HeatMap.h]: Specifying `inline` is unnecessary here.

File: larreco/QuadVtx/HeatMap.h
Comment (knoepfel) [inline larreco/QuadVtx/HeatMap.h:7]: `std::unique_ptr<TH2F>` requires the full type specification of TH2F.  Suggest `#include "TH2F.h"` instead.

File: larreco/QuadVtx/CMakeLists.txt
Comment (knoepfel) [inline larreco/QuadVtx/CMakeLists.txt]: Only modules/services should use services.  The `${ART_FRAMEWORK_SERVICES_REGISTRY}` link line should be removed.  Also, as this library appears to only use facilities from the STL and ROOT, only `${ROOT_*}` libraries should be included in the `LIB_LIBRARIES` list.

File: larreco/QuadVtx/CMakeLists.txt
Comment (knoepfel) [inline larreco/QuadVtx/CMakeLists.txt]: There is no library called `${ART_ROOT_IO_TFILEDIRECTORY}`.  Please remove.

File: larreco/QuadVtx/CMakeLists.txt
Comment (knoepfel) [inline larreco/QuadVtx/CMakeLists.txt]: The `${ROOT_BASIC_LIB_LIST}` is quite large.  Do you really need it all?

File: larreco/QuadVtx/CMakeLists.txt
Comment (knoepfel) [inline larreco/QuadVtx/CMakeLists.txt]: Please remove `BASENAME_ONLY`--LArSoft uses fully-qualified library names.

**Review by knoepfel — CHANGES_REQUESTED** [review]

**Comment by knoepfel** [comment]
@cjbackhouse, thank you for the PR.  I've looked over the code and made many comments.  I'm happy to follow-up on any of them.  Let me know.

File: larreco/QuadVtx/EvalVtx_module.cc
Comment (cjbacchus) [reply larreco/QuadVtx/EvalVtx_module.cc]: Doesn't that mean that protected vs private is completely moot? I generally always use protected because I think you kind of have to trust classes inheriting from you, but I won't fight too hard if protected->private is a requirement.

File: larreco/QuadVtx/CMakeLists.txt
Comment (cjbacchus) [reply larreco/QuadVtx/CMakeLists.txt]: As you can probably tell, the makefile is completely cargo-culted. I'll happily make these changes, or just allow anyone actually knowledgeable to rewrite as part of the pull.

File: larreco/QuadVtx/EvalVtx_module.cc
Comment (cjbacchus) [reply larreco/QuadVtx/EvalVtx_module.cc]: Done

File: larreco/QuadVtx/EvalVtx_module.cc
Comment (cjbacchus) [reply larreco/QuadVtx/EvalVtx_module.cc:131]: The trouble is that the output tree is structured to allow comparison of vertexers event by event. If one vertexer found a vertex and another didn't, this is important information. You can't just drop the whole event. My plotting scripts know what a default-valued vertex means, but it would be possible to add an isValid branch to the tree to make it more explicit, if that's desirable.

File: larreco/QuadVtx/EvalVtx_module.cc
Comment (cjbacchus) [reply larreco/QuadVtx/EvalVtx_module.cc]: Done

File: larreco/QuadVtx/EvalVtx_module.cc
Comment (cjbacchus) [reply larreco/QuadVtx/EvalVtx_module.cc]: Done

File: larreco/QuadVtx/QuadVtx_module.cc
Comment (cjbacchus) [reply larreco/QuadVtx/QuadVtx_module.cc]: Done

File: larreco/QuadVtx/QuadVtx_module.cc
Comment (cjbacchus) [reply larreco/QuadVtx/QuadVtx_module.cc:52]: Done

File: larreco/QuadVtx/QuadVtx_module.cc
Comment (cjbacchus) [reply larreco/QuadVtx/QuadVtx_module.cc]: Done

File: larreco/QuadVtx/QuadVtx_module.cc
Comment (cjbacchus) [reply larreco/QuadVtx/QuadVtx_module.cc]: Done

File: larreco/QuadVtx/QuadVtx_module.cc
Comment (cjbacchus) [reply larreco/QuadVtx/QuadVtx_module.cc]: This made a small but measurable improvement in the speed. Changed to noexcept.

File: larreco/QuadVtx/EvalVtx_module.cc
Comment (cjbacchus) [reply larreco/QuadVtx/EvalVtx_module.cc]: Done

File: larreco/QuadVtx/QuadVtx_module.cc
Comment (cjbacchus) [reply larreco/QuadVtx/QuadVtx_module.cc]: Done, though I suspect it amounts to exactly the same thing for a simple case like this.

File: larreco/QuadVtx/QuadVtx_module.cc
Comment (cjbacchus) [reply larreco/QuadVtx/QuadVtx_module.cc]: Done

File: larreco/QuadVtx/QuadVtx_module.cc
Comment (cjbacchus) [reply larreco/QuadVtx/QuadVtx_module.cc]: Done

File: larreco/QuadVtx/QuadVtx_module.cc
Comment (cjbacchus) [reply larreco/QuadVtx/QuadVtx_module.cc]: Done

File: larreco/QuadVtx/HeatMap.h
Comment (cjbacchus) [reply larreco/QuadVtx/HeatMap.h]: I'm against adding trivial accessors on basic struct-like objects like this. Turns out I can make them const, which should address the main concern.

File: larreco/QuadVtx/HeatMap.h
Comment (cjbacchus) [reply larreco/QuadVtx/HeatMap.h]: Done

File: larreco/QuadVtx/CMakeLists.txt
Comment (cjbacchus) [reply larreco/QuadVtx/CMakeLists.txt]: which variable will give me the libraries for TVectorT et al?

File: larreco/QuadVtx/CMakeLists.txt
Comment (cjbacchus) [reply larreco/QuadVtx/CMakeLists.txt]: Done

File: larreco/QuadVtx/CMakeLists.txt
Comment (cjbacchus) [reply larreco/QuadVtx/CMakeLists.txt]: Done

File: larreco/QuadVtx/QuadVtx_module.cc
Comment (cjbacchus) [reply larreco/QuadVtx/QuadVtx_module.cc]: Done

File: larreco/QuadVtx/EvalVtx_module.cc
Comment (cjbacchus) [reply larreco/QuadVtx/EvalVtx_module.cc]: well, only in the case of multi-threading. But fine, done.

File: larreco/QuadVtx/QuadVtx_module.cc
Comment (cjbacchus) [reply larreco/QuadVtx/QuadVtx_module.cc]: Done

File: larreco/QuadVtx/QuadVtx_module.cc
Comment (cjbacchus) [reply larreco/QuadVtx/QuadVtx_module.cc]: Done

File: larreco/QuadVtx/QuadVtx_module.cc
Comment (cjbacchus) [reply larreco/QuadVtx/QuadVtx_module.cc:298]: __restrict__ does seem to be valid clang syntax, and I got a noticeable speedup from this. If there's a problem, the CI will catch it, right?

File: larreco/QuadVtx/QuadVtx_module.cc
Comment (cjbacchus) [reply larreco/QuadVtx/QuadVtx_module.cc]: Since l goes out of scope on the next line anyway I suspect this generates identical code, but done.

File: larreco/QuadVtx/QuadVtx_module.cc
Comment (cjbacchus) [reply larreco/QuadVtx/QuadVtx_module.cc]: aren't they all within the 'quad' namespace?

File: larreco/QuadVtx/QuadVtx_module.cc
Comment (cjbacchus) [reply larreco/QuadVtx/QuadVtx_module.cc]: Done

File: larreco/QuadVtx/QuadVtx_module.cc
Comment (cjbacchus) [reply larreco/QuadVtx/QuadVtx_module.cc]: Done. This actually got rid of the need for TFileDirectory pointers completely

File: larreco/QuadVtx/HeatMap.h
Comment (cjbacchus) [reply larreco/QuadVtx/HeatMap.h:7]: pretty sure TH2.h explicitly defines all the TH2* types (yes ROOT...), but Done

**Comment by cjbacchus** [comment]
Thanks. For each of your comments I have now either implemented it, or left a comment. Looks like we just need one more round to decide what to do with those.

File: larreco/QuadVtx/EvalVtx_module.cc
Comment (knoepfel) [reply larreco/QuadVtx/EvalVtx_module.cc:131]: I see.  If downstream code already depends on this setup, then it's not worth redoing.  You might consider adding a comment as to why this is being done this way, but that's up to you.  Feel free to resolve this comment according to your liking.

Comment (cjbacchus) [reply larreco/QuadVtx/EvalVtx_module.cc:131]: Added a comment.

File: larreco/QuadVtx/CMakeLists.txt
Comment (knoepfel) [reply larreco/QuadVtx/CMakeLists.txt]: I believe `${ROOT_HIST}` is sufficient for `TVectorT`.

Comment (cjbacchus) [reply larreco/QuadVtx/CMakeLists.txt]: Apparently not, but I fiddled some more and adding Matrix, Physics, and Tree has it building.

File: larreco/QuadVtx/EvalVtx_module.cc
Comment (knoepfel) [reply larreco/QuadVtx/EvalVtx_module.cc]: Sorry, Chris, I didn't give you enough explanation.  A few issues:

- The C++ core guidelines discourage protected data for various reasons (https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rh-protected).  A real-life example where protected data is problematic is the *messagefacility*.   That system has been incredibly difficult to maintain due to its use of protected data: derived classes mess around with base-class state.  We're slowly cleaning it up, but it's difficult to do so.  It's better to use a default protection of `private`, and then to expose data members when it's actually required.
- Even though nothing will (or *should*) inherit from a module, specifying `private` instead of `protected` does not hurt anything, is fewer characters to type, and it keeps with the pattern recommended by the core guidelines.  People don't have to think about why one module uses `protected` data and the other `private`.
- Specifying `protected` in a module could imply to someone down the road that the module was designed to be inherited from.  That's not something we want people to do.  In fact, I sometimes type `class MyModule final : public art::EDProducer {...}`, with the `final` keyword to be even more explicit.

Comment (cjbacchus) [reply larreco/QuadVtx/EvalVtx_module.cc]: OK, done

File: larreco/QuadVtx/QuadVtx_module.cc
Comment (knoepfel) [reply larreco/QuadVtx/QuadVtx_module.cc:298]: Well...let's give it a try, and if we have problems we'll deal with them then.

**Review by knoepfel — APPROVED** [review]

File: larreco/QuadVtx/QuadVtx_module.cc
Comment (knoepfel) [reply larreco/QuadVtx/QuadVtx_module.cc]: Oops, sorry I missed that.  In principle, even though it's in the `quad` namespace, if another library were to define the same function names/signatures in the `quad` namespace, that would be a violation of the one-definition rule.  The run-time is allowed to use the first definition it finds, which might not be what's desired.  To avoid that, it's better to put them in an anonymous namespace, which would make them visible only to this module.

Having said all that, it's pretty unlikely that someone would run into that problem.  It won't hold up the approval.

File: larreco/QuadVtx/HeatMap.h
Comment (knoepfel) [reply larreco/QuadVtx/HeatMap.h]: Ah...yes, this is the frustration about classes vs. structs.  If the type is intended to be struct-like, then keeping them public is the way to go.  That would argue toward changing from `class` to `struct`, as the code is more idiomatic of structs.  But it's your call if you want to do that.

**Comment by knoepfel** [comment]
Chris, thanks for your willingness to go through this review.  We could always find more ways to improve code, but at some point one needs to be practical and move on.  I think the code is now clearer and cleaner.  Cheers. -Kyle

**Comment by knoepfel** [comment]
approve

**Comment by gartung** [comment]
trigger build

**Comment by cjbacchus** [comment]
Thanks!

**Comment by knoepfel** [comment]
trigger build

**Pull request merged**
